skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+                self.bandwidth_cov_ *= gaussian_correction_term
+
+        self.covariance_ = self._cov_lls(
+            self.bandwidth_cov_,
+            self.t_covariance_,
+            self.t_covariance_,
+            raw_cov_coords,
+            raw_cov_values,
+            win,
+        )
+
+        pc_data = self._get_pc(
+            self.covariance_,
+            self.n_components,
+        )
+
+        n_components, explained_variance_ratio_ = pc_data[:2]
+        eigenvalues, phi = pc_data[2:]
+
+        self.n_components = n_components
+        self.explained_variance_ratio_ = explained_variance_ratio_
+        self.explained_variance_ = eigenvalues
+
+        self.components_: FDataGrid = FDataGrid(
+            data_matrix=phi,
+            grid_points=self.mean_.grid_points,
+            domain_range=X.domain_range,
+            dataset_name="Principal Components",
+            argument_names=X.argument_names,
+            coordinate_names=X.coordinate_names,
+            sample_names=[f"Eigenfunction {i+1}" for i in range(phi.shape[0])],
+            extrapolation=X.extrapolation,
+            interpolation=X.interpolation,
+        )
+
+        if self.assume_noisy:
+            self.sigma2_ = self._get_sigma2(
+                self.bandwidth_cov_,
+                self.t_covariance_,
+                raw_cov_coords,
+                raw_cov_values,
+                raw_diag_coords,
+                raw_diag_values,
+                win,
+                np.array(x_work.domain_range),
+            )
+        else:
+            self.sigma2_ = 0
+
+        return self
+
+    def subject_fpc_scores(
+        self,
+        X: FDataIrregular,
+        start: int,
+        end: int,
+        lambda_: NDArrayFloat,
+        t_mean: NDArrayFloat,
+    ) -> NDArrayFloat:
+        """
+        Compute the functional principal component scores for a specific obs.
+
+        Args:
+            X: The functional data object to be analysed.
+            start: The starting index of the subject's data.
+            end: The ending index of the subject's data.
+            lambda_: The eigenvalues of the covariance matrix.
+            t_mean: The time points of the mean function.
+
+        Returns:
+            Principal component scores for the specified subject.
+        """
+        points_i = X.points[start:end].squeeze()
+        values_i = X.values[start:end].squeeze()
+        if points_i.ndim == 0:
+            points_i = np.array([points_i])
+        m_i = len(points_i)
+
+        # Get indices in t_mean_ corresponding to points_i
+        indices = [np.argmin(np.abs(t_mean - pt)) for pt in points_i]
+        mu_i = self.mean_.data_matrix[0, indices].squeeze()
+        phi_i_raw = self.components_.data_matrix[:, indices]
+        phi_i = phi_i_raw[..., 0].T
+
+        num = lambda_ @ phi_i.T
+        denom = phi_i @ lambda_ @ phi_i.T + self.sigma2_ * np.eye(m_i)
+        try:
+            denom_inv = np.linalg.inv(denom)
```

Comment:
Are you sure we *need* this? Most uses of multiplication by an inverse can be replaced with `np.solve`, which gives better results.